# ROMPEBRECHAS
Integrantes:
Mateo Pereyra-Mauro Martines-Raul Porras

In [ ]:
!pip install dbfread

<a name='manejoarchivos'></a>
# <font color=blue> INSPECCIÓN INICIAL E INTEGRACIÓN DE FUENTES</font>

### <font color=green> 1. Importar datos </font>

**PROBLEMA**:
¿Qué factores influyen en casos de violencia escolar a nivel nacional?


**UNIDAD DE ANALÍSIS DE CADA BASE DE DATOS:**
La base de datos de Lineal_1A tiene como unidad de analísis los colegios a nivel nacional. Por otro lado, denuncias escolares tiene como unidad de analísis a cada caso de denuncia o hecho de violencia escolar. Finalmente la base de datos de padrón web tiene como unidad de analísis a cada colegio o institución educativa


**FUENTES DE DATOS USADAS:**

In [ ]:
from dbfread import DBF
import pandas as pd
import numpy as np

RUTA_F1 = "Padron_web.dbf"     # F1 - Padron de IIEE (MINEDU-ESCALE)
RUTA_F2 = "Lineal_3AP_1.dbf"

tabla_f1 = DBF(RUTA_F1, encoding="cp850",   char_decode_errors="replace")
tabla_f2 = DBF(RUTA_F2, encoding="latin-1", char_decode_errors="replace")

df  = pd.DataFrame(iter(tabla_f1))   # F1 - Padron
df1_primaria = pd.DataFrame(iter(tabla_f2))   # F2 - Cedula 3AP

display(df.head(5))
display(df1_primaria.head(5))



,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,GESTION,D_GESTION,GES_DEP,D_GES_DEP,DIRECTOR,TELEFONO,EMAIL,PAGWEB,DIR_CEN,LOCALIDAD,CODCP_INEI,CODCCPP,CEN_POB,AREA_CENSO,DAREACENSO,CODGEO,D_DPTO,D_PROV,D_DIST,D_REGION,CODOOII,D_DREUGEL,NLAT_IE,NLONG_IE,TIPOPROG,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT
0,24953981,0415547,0,016100,123,A2,Inicial - Jardín,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,CACERES DE MAUTINO YONNY ESCOLASTICA,,,,JIRON TERESA GONZALES DE FANNY 543,NICRUPAMPA,0201050001,129688,CENTENARIO,1,Urbana,020105,ANCASH,HUARAZ,INDEPENDENCIA,DRE ANCASH,020001,UGEL HUARAZ,-9.518850,-77.531910,,,11,Mañana,,,,1,Activo,,05-08-2026
1,25273230,0415638,0,015172,122,A2,Inicial - Jardín,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,CACERES DE MAUTINO YONNY ESCOLASTICA,043-420252,,,JIRON 28 DE JULIO S/N,HUARUPAMPA,0201010001,114517,HUARUPAMPA,1,Urbana,020101,ANCASH,HUARAZ,HUARAZ,DRE ANCASH,020001,UGEL HUARAZ,-9.530670,-77.531960,,,11,Mañana,,,,1,Activo,02-05-1942,05-08-2026
2,20414731,0415646,0,015186,233,A2,Inicial - Jardín,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,LOPEZ MONGE MARIA ESTHER MONICA,043-427649 - 943940593,,,JIRON AMADEO FIGUEROA S/N,,,610619,LA SOLEDAD,1,Urbana,020101,ANCASH,HUARAZ,HUARAZ,DRE ANCASH,020001,UGEL HUARAZ,-9.531100,-77.522700,,,13,Mañana-Tarde,,,,1,Activo,04-08-1974,05-08-2026
3,26281962,0415877,0,016751,COLEGIO PARROQUIAL NUESTRA SEÑORA DEL SAGRADO ...,A2,Inicial - Jardín,Escolarizada,3,Mixto,3,Privada,B2,Comunidad o asociación religiosa,DAVILA CALVO ROMMEL SANTIAGO,043-421652,nsscj2005@yahoo.com,,JIRON LOS CAPULIES 163,,0201050001,129688,CENTENARIO,1,Urbana,020105,ANCASH,HUARAZ,INDEPENDENCIA,DRE ANCASH,020001,UGEL HUARAZ,-9.515578,-77.532393,,,13,Mañana-Tarde,20146930418,C.E.P. NUESTRA SEÑORA DEL SAGRADO CORAZON DE J...,,1,Activo,07-05-1962,05-08-2026
4,25111788,0567206,0,016119,268,A2,Inicial - Jardín,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,HUARAC CHAUCA REYNA AURORA,,,,MARIAN,,0201050064,124291,MARIAN,2,Rural,020105,ANCASH,HUARAZ,INDEPENDENCIA,DRE ANCASH,020001,UGEL HUARAZ,-9.513940,-77.504026,,,11,Mañana,,,,1,Activo,24-04-1981,05-08-2026


,cod_mod,anexo,codlocal,nroced,cen_edu,distrito,localidad,p104a,p105a,p105a_s_1,p105a_s_2,p105a_s_3,p105a_s_4,p105a_s_5,p105a_s_5o,p106a,p107a,p107a_no,p107a_no_o,p108a,p109a_1,p109a_2,p109a_3,p109a_4,p109a_5,p109a_6,p110a,p111a_1,p111a_2,p111a_3,p111a_4,p111a_5,p111a_6,p111a_5_ot,p112a,p112a_si,p113a,p113a_s_1,p113a_s_2,p113a_s_3,p113a_s_4,p113a_s_5,p113a_s_6,p113a_s_6o,p114a,p114a_s_1,p114a_s_2,p114a_s_3,p114a_s_4,p114a_s_5,...,p125b,p126b,p126b_si_1,p126b_si_2,p127b,p128b,p128b_si,p129b,p130b,p131b_1,p131b_2,p131b_3,p131b_4,p131b_5,p131b_6,p131b_7,p132b_1,p132b_2,p133b,p133b_si,p134b,p134b_s1_1,p134b_s1_2,p134b_s1_3,p134b_s1_4,p134b_s1_5,p134b_s1_6,p134b_s1_7,p134b_s1_o,p134b_s2_1,p134b_s2_2,p134b_s2_3,p135b,p135b_s_1,p135b_s_2,p135b_s_3,p135b_s_4,p135b_s_5,p135b_si_5,p136b,p136b_s_1,p136b_s_2,p136b_s_3,p136b_s_4,p136b_s_4o,niv_mod,ges_dep,area_censo,codgeo,codooii
0,1720556,0,852373,3AP,SIR ALEXANDER FLEMING,CERCADO,,NO,,,,,,,,NO,,,,NO,X,,,,,,SI,X,,,X,,,,NO,,NS,,,,,,,,NO,,,,,,...,SI,NO,,//,SI,SI,SI,SI,SI,X,X,X,,X,,,,,SI,SI,NO,,,,,,,,,,,,SI,X,,,,,,SI,X,X,,,,B0,B4,1,040101,040001
1,0313890,0,073181,3AP,40513 SAN MIGUEL ARCANGEL,ALCA,ANEXO DE CAHUANA,NO,,,,,,,,NO,,,,SI,,,,,,X,SI,X,X,X,X,,,,SI,SI,SI,,,X,,,,,SI,X,,X,,,...,SI,NO,,//,NO,SI,SI,SI,SI,X,X,X,X,X,,,31,58,SI,SI,SI,X,X,X,,,,,,X,,X,SI,X,X,X,X,,,SI,X,X,X,,,B0,A1,2,040802,040009
2,1495092,0,605028,3AP,33421,PILLCO MARCA,CAYHUAYNA,NO,,,,,,,,SI,NO,2,,SI,,,,X,,,SI,X,,X,,,,,SI,SI,NO,,,,,,,,SI,,X,,,,...,SI,SI,1,25/7/2024,SI,SI,NO,SI,SI,X,X,,,,,,,,,,NO,,,,,,,,,,,,NO,,,,,,,NO,,,,,,B0,A1,1,100111,100001
3,1579861,0,651751,3AP,SANTANDER DE PUENTE PIEDRA,PUENTE PIEDRA,SANTA PAULA,NO,,,,,,,,SI,SI,,,SI,,,,,,X,SI,X,,,,,,,NO,,NO,,,,,,,,SI,,,X,,,...,SI,NO,,//,NO,NO,,SI,SI,,X,,,,,,,,,,NO,,,,,,,,,,,,NO,,,,,,,NO,,,,,,B0,B4,1,150125,150105
4,1529767,0,553166,3AP,COLEGIO SUPERIOR DE CIENCIAS NEWTON VIANNEY,YURA,,NO,,,,,,,,SI,SI,,,SI,,,,,,X,SI,X,,,,,,,NO,,NO,,,,,,,,NO,,,,,,...,SI,NO,,//,NO,NO,,SI,NO,,,X,,,,,,OFICIO 013,SI,SI,NO,,,,,,,,,,,,NO,,,,,,,SI,,X,,,,B0,B4,1,040128,040001


In [ ]:
import pandas as pd
# Cargamos el archivo saltando las filas de metadatos iniciales
# header=5 indica que los nombres de las columnas están en la quinta fila del archivo Excel
df2 = pd.read_excel("denuncias_escolares.xlsx", header=5)
# Visualizamos el resultado limpio
display(df2.head())



,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
0,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,Hostigamiento sexual,Atención en proceso
1,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención en proceso
2,2026-07-31,DRE Arequipa,UGEL Arequipa Sur,Secundaria,Entre Escolares,Psicológica,Verbal,Pendiente de atención por la IE
3,2026-07-31,DRE Ica,UGEL Ica,Secundaria,Entre Escolares,Física,Sin lesiones,Pendiente de atención por la IE
4,2026-07-31,DRE Cusco,UGEL Anta,Secundaria,Entre Escolares,Psicológica,Con uso de armas,Pendiente de atención por la IE


### <font color=green> 2. REVISAR EL TAMAÑO DE CADA FUENTE </font>

In [ ]:
display(df.columns)
display(df1.columns)
display(df2.columns)

print("="*500)

print("Tamaño de fuente 1:", df.shape)
print("Tamaño de fuente 2:", df1.shape)
print("Tamaño de fuente 3:", df2.shape)

Index(['CODINST', 'COD_MOD', 'ANEXO', 'CODLOCAL', 'CEN_EDU', 'NIV_MOD',
       'D_NIV_MOD', 'D_FORMA', 'TIPSSEXO', 'D_TIPSSEXO', 'GESTION',
       'D_GESTION', 'GES_DEP', 'D_GES_DEP', 'DIRECTOR', 'TELEFONO', 'EMAIL',
       'PAGWEB', 'DIR_CEN', 'LOCALIDAD', 'CODCP_INEI', 'CODCCPP', 'CEN_POB',
       'AREA_CENSO', 'DAREACENSO', 'CODGEO', 'D_DPTO', 'D_PROV', 'D_DIST',
       'D_REGION', 'CODOOII', 'D_DREUGEL', 'NLAT_IE', 'NLONG_IE', 'TIPOPROG',
       'D_TIPOPROG', 'COD_TUR', 'D_COD_TUR', 'NRORUC', 'RZSOCIAL', 'PROMOTOR',
       'ESTADO', 'D_ESTADO', 'FECHAREG', 'FECHA_ACT'],
      dtype='object')

Index(['COD_MOD', 'ANEXO', 'CODLOCAL', 'NROCED', 'CEN_EDU', 'DISTRITO',
       'LOCALIDAD', 'ALUM_35IIC', 'ALUM_02IC', 'PRO_CONVER',
       ...
       'P110C_S_3O', 'P111C', 'P112C_1', 'P112C_2', 'P112C_3', 'NIV_MOD',
       'GES_DEP', 'AREA_CENSO', 'CODGEO', 'CODOOII'],
      dtype='object', length=230)

Index(['FECHA_REPORTE', 'DRE', 'UGEL', 'NIVEL_EDUCATIVO', 'TIPO_REPORTE',
       'TIPO_VIOLENCIA', 'SUBTIPO_VIOLENCIA', 'TIPO_ESTADO_REPORTE'],
      dtype='object')

Tamaño de fuente 1: (56858, 45)
Tamaño de fuente 2: (34097, 230)
Tamaño de fuente 3: (48273, 8)


### <font color=green> 3. REVISAR LOS TIPOS DE DATOS DETECTADOS </font>


In [ ]:
display(df.sample(5))
display(df1.sample(5))
display(df2.sample(5))

,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,GESTION,D_GESTION,GES_DEP,D_GES_DEP,DIRECTOR,TELEFONO,EMAIL,PAGWEB,DIR_CEN,LOCALIDAD,CODCP_INEI,CODCCPP,CEN_POB,AREA_CENSO,DAREACENSO,CODGEO,D_DPTO,D_PROV,D_DIST,D_REGION,CODOOII,D_DREUGEL,NLAT_IE,NLONG_IE,TIPOPROG,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT
56230,26844967,1474303,0,687161,DIVINO CREADOR,B0,Primaria,Escolarizada,3,Mixto,3,Privada,B4,Particular,SANTA CRUZ BECERRA ILMA,2834229,,,AVENIDA JOSE CARLOS MARIATEGUI 1245,JOSE CARLOS MARIATEGUI,1501430001,546446,VILLA MARIA DEL TRIUNFO,1,Urbana,150143,LIMA,LIMA,VILLA MARIA DEL TRIUNFO,DRE LIMA METROPOLITANA,150102,UGEL 01 SAN JUAN DE MIRAFLORES,-12.151780,-76.951580,,,11,Mañana,,,ILMA SANTA CRUZ BECERRA,1,Activo,22-05-2002,05-08-2026
33559,,1540574,0,753518,CURINGA,L9,Educación Ocupacional,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,RIO BLANCO,,1605080001,130445,CURINGA,2,Rural,160508,LORETO,REQUENA,SOPLIN,DRE LORETO,160006,UGEL REQUENA,-6.007986,-73.692750,,,11,Mañana,,,,2,Inactivo,,05-08-2026
50068,26253251,1388677,0,150952,50118 JOSE MARIA ARGUEDAS,F0,Secundaria,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,BACA GAMARRA WILY TOMAS,,,,PANTIPATA,,0803040027,124389,PANTIPATA,2,Rural,080304,CUSCO,ANTA,CHINCHAYPUJIO,DRE CUSCO,080003,UGEL ANTA,-13.582348,-72.334979,,,11,Mañana,,,,1,Activo,,05-08-2026
34368,24131696,0386987,0,466051,72445 GLORIOSO 835 SAN JUAN BOSCO,B0,Primaria,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,PAREDES ARIZAPANA BLASCO WILBERDO,,,,LIMBANI,,2112030001,116081,LIMBANI,2,Rural,211203,PUNO,SANDIA,LIMBANI,DRE PUNO,210014,UGEL CRUCERO,-14.152238,-69.689815,,,11,Mañana,,,,1,Activo,,05-08-2026
50492,25864514,0234237,0,158670,56305,B0,Primaria,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,MENDOZA PINTO BERNARDINO FELIBERTO,,,,AYACCASI,,0807080008,119487,AYACCASI,2,Rural,080708,CUSCO,CHUMBIVILCAS,VELILLE,DRE CUSCO,080007,UGEL CHUMBIVILCAS,-14.415536,-71.888649,,,11,Mañana,,,,1,Activo,30-04-1971,05-08-2026


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,ALUM_35IIC,ALUM_02IC,PRO_CONVER,ANIOCONVER,P104A,P105A,P105A_S_1,P105A_S_2,P105A_S_3,P105A_S_4,P105A_S_5,P105A_S_5O,P106A,P107A,P107A_NO,P107A_NO_O,P108A,P109A_1,P109A_2,P109A_3,P109A_4,P110A,P111A_1,P111A_2,P111A_3,P111A_4,P111A_5,P111A_6,P112A,P113A_1,P113A_2,P113A_3,P113A_4,P113A_5,P113A_6,P113A_5_OT,P114A,P114A_SI,P115A,P115A_S_1,P115A_S_2,P115A_S_3,P115A_S_4,...,P103C_SI_1,P103C_SI_2,P103C_SI_3,P103C_SI_4,P103C_SI_A,P103C_SI_B,P104C_1,P104C_2,P104C_3,P104C_4,P104C_5,P104C_6,P104C_7,P104C_8,P104C_10,P104C_10_O,P105C,P105C_SI,P106C,P106C_4_OT,P107C,P108C,P108C_SI_1,P108C_SI_2,P108C_SI_3,P108C_SI_4,P108C_SI_5,P108C_SI_6,P108C_SI_7,P108C_SI_8,P109C,P109C_S_1,P109C_S_2,P109C_S_3,P109C_S_4,P109C_S_4O,P110C,P110C_S_1,P110C_S_2,P110C_S_3,P110C_S_3O,P111C,P112C_1,P112C_2,P112C_3,NIV_MOD,GES_DEP,AREA_CENSO,CODGEO,CODOOII
19733,0259671,0,413152,1A,003,CATACAOS,CATACAOS,,NO,NO,,SI,SI,,,,,X,CLASES VIRTUALES,SI,NO,1,,NO,X,,X,,SI,,,X,,X,,SI,X,X,X,,X,,PROGRAMA WASI MIKUNA,NO,,SI,,,X,,...,,X,,,,,X,,,X,X,,,X,,,SI,3,1,,SI,SI,,X,X,X,X,,,,SI,X,,X,,,SI,X,,,,SI,X,,,A2,A1,1,200105,200001
33173,1785633,0,853264,1A,790-B,IPARIA,PUERTO ESPERANZA DE SHESHEA,,NO,NO,,NO,,,,,,,,SI,SI,,,NO,X,,,,SI,X,,X,,,,NO,,,,,,X,,NO,,NS,,,,,...,,,,,,,,,,,,,,,,,SI,3,1,,NO,NO,,,,,,,,,NO,,,,,,,,,,,SI,X,X,,A2,A1,2,250103,250001
32497,0876516,0,332151,1A,0082 LA CANTUTA,SAN LUIS,URB EL PINO,,NO,NO,,NO,,,,,,,,SI,SI,,,NO,X,,,,SI,X,,,X,X,,SI,X,,X,,X,,TAMIZAJE,SI,SI,SI,,,X,,...,,X,,,,,X,X,X,X,X,,X,X,,,SI,2,1,,SI,SI,,,X,,,X,,,SI,X,,X,,,SI,X,X,,,SI,,X,,A2,A1,1,150134,150108
25756,1682046,0,767539,1A,1211,HUAMANCACA CHICO,SANTA R5OSA,,NO,NO,,NO,,,,,,,,SI,SI,,,SI,X,,X,,SI,X,,,,,,SI,,,,,,X,,NO,,NO,,,,,...,,X,,,,,X,,,,,,X,,X,DISTRIBUCION DE LOS ESPACIOS EDUCATIVOS,SI,2,1,,NO,SI,,,,X,,,,,SI,X,,,,,SI,X,,,,SI,X,,,A2,A1,1,120905,120002
15937,1202084,0,149379,1A,NUESTRA SEÑORA VIRGEN DEL CARMEN,WANCHAQ,WANCHAQ,,NO,NO,,NO,,,,,,,,SI,SI,,,NO,X,X,X,,SI,,,,,,X,SI,X,,,,,,,NO,,NO,,,,,...,,,,,,,,,,,,,,,,,SI,2,3,,NO,NO,,,,,,,,,SI,X,,X,,,SI,,,X,TALLERES,SI,,X,,A2,B4,1,080108,080001


,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
46326,2024-04-18,DRE Junín,UGEL Huancayo,Primaria,Personal IE a Escolares,Física,Castigo físico,Atención finalizada
8134,2026-04-08,DRE Puno,UGEL Puno,Secundaria,Entre Escolares,Física,Sin lesiones,Atención finalizada
4737,2026-05-27,DRE Piura,UGEL La Unión,Secundaria,Entre Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención en proceso
16854,2025-09-17,DRE Amazonas,UGEL Intercultural Bilingüe Cenepa,Secundaria,Personal IE a Escolares,Sexual,Violación sexual,Atención finalizada
34137,2024-10-14,DRE Lima Metropolitana,UGEL 03 Cercado,Secundaria,Personal IE a Escolares,Psicológica,Trato humillante,Atención finalizada


In [ ]:
display(df.dtypes)
display(df1.dtypes)
display(df2.dtypes)

,0
CODINST,object
COD_MOD,object
ANEXO,object
CODLOCAL,object
CEN_EDU,object
NIV_MOD,object
D_NIV_MOD,object
D_FORMA,object
TIPSSEXO,object
D_TIPSSEXO,object


,0
COD_MOD,object
ANEXO,object
CODLOCAL,object
NROCED,object
CEN_EDU,object
...,...
NIV_MOD,object
GES_DEP,object
AREA_CENSO,object
CODGEO,object


,0
FECHA_REPORTE,datetime64[ns]
DRE,object
UGEL,object
NIVEL_EDUCATIVO,object
TIPO_REPORTE,object
TIPO_VIOLENCIA,object
SUBTIPO_VIOLENCIA,object
TIPO_ESTADO_REPORTE,object


In [ ]:
display(df1.columns)

Index(['COD_MOD', 'ANEXO', 'CODLOCAL', 'NROCED', 'CEN_EDU', 'DISTRITO',
       'LOCALIDAD', 'ALUM_35IIC', 'ALUM_02IC', 'PRO_CONVER',
       ...
       'P110C_S_3O', 'P111C', 'P112C_1', 'P112C_2', 'P112C_3', 'NIV_MOD',
       'GES_DEP', 'AREA_CENSO', 'CODGEO', 'CODOOII'],
      dtype='object', length=230)

In [ ]:
print(f"el dataframe tiene {df.shape} filas y columnas respectivamente")
print(f"el dataframe tiene {df1.shape} filas y columnas respectivamente")
print(f"el dataframe tiene {df2.shape} filas y columnas respectivamente")

el dataframe tiene (56858, 45) filas y columnas respectivamente
el dataframe tiene (34097, 230) filas y columnas respectivamente
el dataframe tiene (48273, 8) filas y columnas respectivamente


### <font color=green> 4. USAR describe() PARA RESUMIR CADA FUENTE </font>

In [ ]:
# describe(include="all") resume variables numéricas y categóricas.

print("Resumen de fuente 1:")
display(df.describe(include="all"))
print("Resumen de fuente 2:")
display(df1.describe(include="all"))
print("Resumen de fuente 3:")
display(df2.describe(include="all"))

Resumen de fuente 1:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,GESTION,D_GESTION,GES_DEP,D_GES_DEP,DIRECTOR,TELEFONO,EMAIL,PAGWEB,DIR_CEN,LOCALIDAD,CODCP_INEI,CODCCPP,CEN_POB,AREA_CENSO,DAREACENSO,CODGEO,D_DPTO,D_PROV,D_DIST,D_REGION,CODOOII,D_DREUGEL,NLAT_IE,NLONG_IE,TIPOPROG,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT
count,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56857.000000,56857.000000,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858,56858
unique,36194,56301,8,46635,35825,27,19,3,3,3,3,3,10,10,43548,9818,3046,533,32039,4958,21268,23240,16964,3,3,1746,26,195,1613,27,231,231,NaN,NaN,1,1,8,8,1162,1364,2410,3,3,10376,2
top,,0565564,0,658446,JOSE CARLOS MARIATEGUI,B0,Primaria,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,PLAZA PRINCIPAL,,,535994,TRUJILLO,2,Rural,130101,CAJAMARCA,LIMA,TRUJILLO,DRE CAJAMARCA,150102,UGEL 01 SAN JUAN DE MIRAFLORES,NaN,NaN,,,11,Mañana,,,,1,Activo,,05-08-2026
freq,13566,8,56278,9,143,29036,29036,55067,55817,55817,41070,41070,40785,40785,2588,41764,52751,56040,1002,40798,4285,908,871,32321,32321,926,5684,2944,926,5684,2474,2474,NaN,NaN,56858,56858,47682,47682,54422,54080,52680,44368,44368,14133,56857
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-10.500481,-75.067889,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.786436,6.391730,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-18.334575,-81.278494,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-13.516702,-77.779785,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-11.002735,-76.139930,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-7.278410,-73.262770,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Resumen de fuente 2:


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,ALUM_35IIC,ALUM_02IC,PRO_CONVER,ANIOCONVER,P104A,P105A,P105A_S_1,P105A_S_2,P105A_S_3,P105A_S_4,P105A_S_5,P105A_S_5O,P106A,P107A,P107A_NO,P107A_NO_O,P108A,P109A_1,P109A_2,P109A_3,P109A_4,P110A,P111A_1,P111A_2,P111A_3,P111A_4,P111A_5,P111A_6,P112A,P113A_1,P113A_2,P113A_3,P113A_4,P113A_5,P113A_6,P113A_5_OT,P114A,P114A_SI,P115A,P115A_S_1,P115A_S_2,P115A_S_3,P115A_S_4,...,P103C_SI_1,P103C_SI_2,P103C_SI_3,P103C_SI_4,P103C_SI_A,P103C_SI_B,P104C_1,P104C_2,P104C_3,P104C_4,P104C_5,P104C_6,P104C_7,P104C_8,P104C_10,P104C_10_O,P105C,P105C_SI,P106C,P106C_4_OT,P107C,P108C,P108C_SI_1,P108C_SI_2,P108C_SI_3,P108C_SI_4,P108C_SI_5,P108C_SI_6,P108C_SI_7,P108C_SI_8,P109C,P109C_S_1,P109C_S_2,P109C_S_3,P109C_S_4,P109C_S_4O,P110C,P110C_S_1,P110C_S_2,P110C_S_3,P110C_S_3O,P111C,P112C_1,P112C_2,P112C_3,NIV_MOD,GES_DEP,AREA_CENSO,CODGEO,CODOOII
count,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,...,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097
unique,34097,1,34091,1,19704,2795,16209,3,3,3,69,3,3,2,2,2,2,2,169,3,3,5,222,3,3,2,2,2,3,2,2,2,2,2,2,3,2,2,2,2,2,2,1272,3,3,4,2,2,2,2,...,3,2,3,3,226,323,3,3,3,3,3,3,3,3,3,324,3,6,5,221,3,3,3,3,2,2,3,2,2,3,3,2,2,2,3,398,3,2,2,3,1111,3,3,3,3,3,10,2,1882,225
top,1643303,0,530151,1A,INICIAL,SAN JUAN DE LURIGANCHO,,,NO,NO,,NO,,,,,,,,SI,SI,,,NO,X,,,,SI,X,,,,,,SI,X,,,,,,,NO,,SI,,,,,...,,X,,,,,X,,,,,,,,,,SI,2,1,,NO,SI,,,,,,,,,SI,,,,,,SI,,,,,SI,,X,,A2,A1,2,150132,150102
freq,1,34097,2,34097,67,507,1300,33633,30180,27220,27202,31861,30890,33157,33681,33972,33708,33621,33616,21653,21216,32663,33839,17352,28536,26237,17685,30434,32338,21358,30838,31738,29029,26476,24181,33590,19276,21239,29292,31646,31202,25026,31497,21584,21091,16345,32629,33103,19629,32713,...,32259,17325,33639,33540,33637,33537,17132,25783,25022,26089,22607,29202,24966,25810,33626,33627,33525,18633,24983,33684,21762,21405,29043,24019,27438,26752,30015,27223,28483,32001,20630,20425,29383,21804,33382,33384,20840,20771,24814,31304,31304,26537,20345,18794,26360,32343,24593,18078,533,940


Resumen de fuente 3:


,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
count,48273,48273,48273,48273,48273,48273,48273,48273
unique,NaN,26,225,8,2,3,17,5
top,NaN,DRE Lima Metropolitana,UGEL 03 Cercado,Secundaria,Entre Escolares,Física,Sin lesiones,Atención finalizada
freq,NaN,16217,3344,30653,31098,20408,9839,37637
mean,2025-05-11 16:42:43.022807552,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,2024-01-02 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,2024-09-17 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,2025-06-06 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,2025-11-19 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN
max,2026-07-31 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### <font color=green> 5. REVISAR VALORES FALTANTES </font>

In [ ]:
print("Valores faltantes en fuente1:")
display(df.isnull().sum())


print("Porcentaje de valores faltantes en fuente1:")
display((df.isnull().mean() * 100).round(2))

print("="*200)

print("Valores faltantes en fuente2:")
display(df1.isnull().sum())


print("Porcentaje de valores faltantes en fuente2:")
display((df1.isnull().mean() * 100).round(2))


Valores faltantes en fuente1:


,0
CODINST,0
COD_MOD,0
ANEXO,0
CODLOCAL,0
CEN_EDU,0
NIV_MOD,0
D_NIV_MOD,0
D_FORMA,0
TIPSSEXO,0
D_TIPSSEXO,0


Porcentaje de valores faltantes en fuente1:


,0
CODINST,0.0
COD_MOD,0.0
ANEXO,0.0
CODLOCAL,0.0
CEN_EDU,0.0
NIV_MOD,0.0
D_NIV_MOD,0.0
D_FORMA,0.0
TIPSSEXO,0.0
D_TIPSSEXO,0.0


Valores faltantes en fuente2:


,0
COD_MOD,0
ANEXO,0
CODLOCAL,0
NROCED,0
CEN_EDU,0
...,...
NIV_MOD,0
GES_DEP,0
AREA_CENSO,0
CODGEO,0


Porcentaje de valores faltantes en fuente2:


,0
COD_MOD,0.0
ANEXO,0.0
CODLOCAL,0.0
NROCED,0.0
CEN_EDU,0.0
...,...
NIV_MOD,0.0
GES_DEP,0.0
AREA_CENSO,0.0
CODGEO,0.0


In [ ]:
print("Valores faltantes en fuente3:")
display(df2.isnull().sum())


print("Porcentaje de valores faltantes en fuente3:")
display((df2.isnull().mean() * 100).round(2))

Valores faltantes en fuente3:


,0
FECHA_REPORTE,0
DRE,0
UGEL,0
NIVEL_EDUCATIVO,0
TIPO_REPORTE,0
TIPO_VIOLENCIA,0
SUBTIPO_VIOLENCIA,0
TIPO_ESTADO_REPORTE,0


Porcentaje de valores faltantes en fuente3:


,0
FECHA_REPORTE,0.0
DRE,0.0
UGEL,0.0
NIVEL_EDUCATIVO,0.0
TIPO_REPORTE,0.0
TIPO_VIOLENCIA,0.0
SUBTIPO_VIOLENCIA,0.0
TIPO_ESTADO_REPORTE,0.0


### <font color=green> 6. REVISAR CANTIDAD DE VALORES DISTINTOS POR VARIABLE

In [ ]:

print("Valores distintos por variable en fuente 1:")
display(df.nunique())

print("Valores distintos por variable en fuente 2:")
display(df1.nunique())

print("Valores distintos por variable en fuente 3:")
display(df2.nunique())

Valores distintos por variable en fuente 1:


,0
CODINST,68235
COD_MOD,179383
ANEXO,26
CODLOCAL,105539
CEN_EDU,81916
NIV_MOD,29
D_NIV_MOD,21
D_FORMA,3
TIPSSEXO,3
D_TIPSSEXO,3


Valores distintos por variable en fuente 2:


,0
COD_MOD,34097
ANEXO,1
CODLOCAL,34091
NROCED,1
CEN_EDU,19704
...,...
NIV_MOD,3
GES_DEP,10
AREA_CENSO,2
CODGEO,1882


Valores distintos por variable en fuente 3:


,0
FECHA_REPORTE,892
DRE,26
UGEL,225
NIVEL_EDUCATIVO,8
TIPO_REPORTE,2
TIPO_VIOLENCIA,3
SUBTIPO_VIOLENCIA,17
TIPO_ESTADO_REPORTE,5


### <font color=green> 7. REVISAR LOS VALORES DE UNA VARIABLE CATEGÓRICA </font>

In [ ]:

print("Valores de la variable sexo en fuente 1:")
display(df["D_TIPSSEXO"].value_counts(dropna=False))

print("Valores de la variable distrito en fuente 2:")
display(df1["DISTRITO"].value_counts(dropna=False))

print("Valores de la variable distrito en fuente 2:")
display(df2["NIVEL_EDUCATIVO"].value_counts(dropna=False))

Valores de la variable sexo en fuente 1:


,count
D_TIPSSEXO,
Mixto,178734
Mujeres,1371
Varones,687


Valores de la variable distrito en fuente 2:


,count
DISTRITO,
SAN JUAN DE LURIGANCHO,507
ATE,384
SAN MARTIN DE PORRES,350
COMAS,285
JULIACA,248
...,...
LA CONVENCION,1
PACARPATA,1
JOSE ELEONARDO ORTIZ,1


Valores de la variable distrito en fuente 2:


,count
NIVEL_EDUCATIVO,
Secundaria,30653
Primaria,15023
Inicial - Jardín,2057
Inicial - Cuna-Jardín,302
Básica Alternativa - Avanzado,127
Básica Especial - Primaria,86
Básica Especial - Inicial,16
Básica Alternativa - Inicial e Intermedio,9


### <font color=green> 8. CREAR UNA TABLA RESUMEN SIMPLE PARA CADA FUENTE </font>

In [ ]:
def resumen_fuente(datos, nombre_fuente):
    """Tabla resumen por variable. Un valor se cuenta como faltante si es NaN
    o si es una cadena vacia (que es como el formato DBF codifica el vacio)."""
    vacio = datos.isna() | (datos.astype(str).apply(lambda s: s.str.strip()) == "")
    return pd.DataFrame({
        "fuente": nombre_fuente,
        "variable": datos.columns,
        "tipo_python": datos.dtypes.astype(str).values,
        "n_faltantes": vacio.sum().values,
        "porcentaje_faltantes": (vacio.mean() * 100).round(2).values,
        "valores_distintos": datos.nunique(dropna=True).values,
    })

resumen_f1 = resumen_fuente(df,  "F1_Padron_web")
resumen_f2 = resumen_fuente(df1, "F2_Lineal_3AP_1")
resumen_f3 = resumen_fuente(df2, "DenunciasEscolares")

tabla_resumen_fuentes = pd.concat([resumen_f1, resumen_f2, resumen_f3], ignore_index=True)

print("F1 - Padron_web (45 variables)")
display(resumen_f1)
print("F2 - Lineal_3AP_1 (230 variables) - primeras 30")
display(resumen_f2.head(30))
print("denuncias_escolares.xlsx ( 8 variables)")
display(resumen_f3)
print("Tabla resumen combinada:", tabla_resumen_fuentes.shape)

F1 - Padron_web (45 variables)


,fuente,variable,tipo_python,n_faltantes,porcentaje_faltantes,valores_distintos
0,F1_Padron_web,CODINST,object,85232,47.14,68235
1,F1_Padron_web,COD_MOD,object,0,0.00,179383
2,F1_Padron_web,ANEXO,object,0,0.00,26
3,F1_Padron_web,CODLOCAL,object,39696,21.96,105539
4,F1_Padron_web,CEN_EDU,object,0,0.00,81916
5,F1_Padron_web,NIV_MOD,object,0,0.00,29
6,F1_Padron_web,D_NIV_MOD,object,0,0.00,21
7,F1_Padron_web,D_FORMA,object,0,0.00,3
8,F1_Padron_web,TIPSSEXO,object,0,0.00,3
9,F1_Padron_web,D_TIPSSEXO,object,0,0.00,3


F2 - Lineal_1A_1 (230 variables) - primeras 30


,fuente,variable,tipo_python,n_faltantes,porcentaje_faltantes,valores_distintos
0,F2_Lineal_1A_1,COD_MOD,object,0,0.00,34097
1,F2_Lineal_1A_1,ANEXO,object,0,0.00,1
2,F2_Lineal_1A_1,CODLOCAL,object,0,0.00,34091
3,F2_Lineal_1A_1,NROCED,object,0,0.00,1
4,F2_Lineal_1A_1,CEN_EDU,object,0,0.00,19704
5,F2_Lineal_1A_1,DISTRITO,object,0,0.00,2795
6,F2_Lineal_1A_1,LOCALIDAD,object,1300,3.81,16209
7,F2_Lineal_1A_1,ALUM_35IIC,object,33633,98.64,3
8,F2_Lineal_1A_1,ALUM_02IC,object,1725,5.06,3
9,F2_Lineal_1A_1,PRO_CONVER,object,36,0.11,3


denuncias_escolares.xlsx ( 8 variables)


,fuente,variable,tipo_python,n_faltantes,porcentaje_faltantes,valores_distintos
0,DenunciasEscolares,FECHA_REPORTE,datetime64[ns],0,0.0,892
1,DenunciasEscolares,DRE,object,0,0.0,26
2,DenunciasEscolares,UGEL,object,0,0.0,225
3,DenunciasEscolares,NIVEL_EDUCATIVO,object,0,0.0,8
4,DenunciasEscolares,TIPO_REPORTE,object,0,0.0,2
5,DenunciasEscolares,TIPO_VIOLENCIA,object,0,0.0,3
6,DenunciasEscolares,SUBTIPO_VIOLENCIA,object,0,0.0,17
7,DenunciasEscolares,TIPO_ESTADO_REPORTE,object,0,0.0,5


Tabla resumen combinada: (283, 6)


### <font color=green> 9. IDENTIFICAR COLUMNAS COMUNES ENTRE LAS FUENTES </font>

In [ ]:
columnas_comunes = sorted(set(df.columns) & set(df1.columns))

print("Columnas con el mismo nombre en ambas fuentes (%d):" % len(columnas_comunes))
print(columnas_comunes)

# No toda columna comun sirve como llave. Se verifica cual es realmente unica.
print("\n--- Unicidad de las llaves candidatas ---")
for nombre, llave in [("COD_MOD solo", ["COD_MOD"]),
                      ("COD_MOD + ANEXO", ["COD_MOD", "ANEXO"])]:
    print("%-18s | F1 duplicados: %6d | F2 duplicados: %6d"
          % (nombre, df.duplicated(llave).sum(), df1.duplicated(llave).sum()))

print("\nOJO: ANEXO es constante en F2 ->", sorted(df1["ANEXO"].unique().tolist()),
      "| en F1 toma", df["ANEXO"].nunique(), "valores")

# Cobertura de la llave compuesta
k1 = set(df["COD_MOD"] + "|" + df["ANEXO"])
k2 = set(df1["COD_MOD"] + "|" + df1["ANEXO"])
print("\nRegistros de F2 que encuentran par en F1: %d de %d (%.1f%%)"
      % (len(k1 & k2), len(k2), len(k1 & k2) / len(k2) * 100))
print("Registros de F2 huerfanos (sin par en F1):", len(k2 - k1))

print("\nDistritos (CODGEO): F1 =", df["CODGEO"].nunique(),
      "| F2 =", df1["CODGEO"].nunique(),
      "| de F2 ausentes en F1 =", len(set(df1["CODGEO"]) - set(df["CODGEO"])))

Columnas con el mismo nombre en ambas fuentes (10):
['ANEXO', 'AREA_CENSO', 'CEN_EDU', 'CODGEO', 'CODLOCAL', 'CODOOII', 'COD_MOD', 'GES_DEP', 'LOCALIDAD', 'NIV_MOD']

--- Unicidad de las llaves candidatas ---
COD_MOD solo       | F1 duplicados:   1409 | F2 duplicados:      0
COD_MOD + ANEXO    | F1 duplicados:      0 | F2 duplicados:      0

OJO: ANEXO es constante en F2 -> ['0'] | en F1 toma 26 valores

Registros de F2 que encuentran par en F1: 34097 de 34097 (100.0%)
Registros de F2 huerfanos (sin par en F1): 0

Distritos (CODGEO): F1 = 1892 | F2 = 1882 | de F2 ausentes en F1 = 0


In [ ]:
# Buscamos columnas comunes entre Lineal_3AP_1.dbf y denuncias escolares.

columnas_comunes = set(df1.columns).intersection(set(df2.columns))

print("Columnas comunes entre las fuentes:")
print(columnas_comunes)


Columnas comunes entre las fuentes:
set()


In [ ]:
# Buscamos columnas comunes entre Lineal_3AP_1.dbf y denuncias escolares.

columnas_comunes = set(df.columns).intersection(set(df2.columns))

print("Columnas comunes entre las fuentes:")
print(columnas_comunes)


Columnas comunes entre las fuentes:
set()


La data de Padron_web tiene una columna (D_DREUGEL-Nombre de la DRE o UGEL que supervisa el servicio educativo) similar a la columna UGEL de denuncias escolares (UGEL) pero no tienen el mismo nombre.

### <font color=green> 10. INTEGRAR LAS DOS FUENTES </font>

In [ ]:
# Llave de integracion: COD_MOD + ANEXO (NO COD_MOD solo: hay 1,409 codigos
# repetidos en F1 y ANEXO es constante '0' en F2, asi que unir por COD_MOD
# perderia o duplicaria silenciosamente los servicios con anexo distinto de 0).

base_integrada = pd.merge(
    df,
    df1,
    on=["COD_MOD", "ANEXO"],
    how="outer",
    suffixes=("_f1", "_f2"),
    indicator=True,
)

print("Base integrada:", base_integrada.shape)
display(base_integrada["_merge"].value_counts())

# Version util para el analisis: solo Inicial escolarizada activa, que es
# donde F1 y F2 se solapan de verdad.
inicial = base_integrada[
    (base_integrada["D_ESTADO"] == "Activo")
    & (base_integrada["NIV_MOD_f1"].isin(["A1", "A2", "A3"]))
].copy()

print("\nInicial escolarizada activa:", len(inicial))
print("...con datos de la Cedula 1A: %d (%.1f%% de cobertura del censo)"
      % ((inicial["_merge"] == "both").sum(),
         (inicial["_merge"] == "both").mean() * 100))

display(base_integrada.head())

Base integrada: (180792, 274)


,count
_merge,
left_only,146695
both,34097
right_only,0



Inicial escolarizada activa: 36127
...con datos de la Cedula 1A: 34015 (94.2% de cobertura del censo)


,CODINST,COD_MOD,ANEXO,CODLOCAL_f1,CEN_EDU_f1,NIV_MOD_f1,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,GESTION,D_GESTION,GES_DEP_f1,D_GES_DEP,DIRECTOR,TELEFONO,EMAIL,PAGWEB,DIR_CEN,LOCALIDAD_f1,CODCP_INEI,CODCCPP,CEN_POB,AREA_CENSO_f1,DAREACENSO,CODGEO_f1,D_DPTO,D_PROV,D_DIST,D_REGION,CODOOII_f1,D_DREUGEL,NLAT_IE,NLONG_IE,TIPOPROG,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,CODLOCAL_f2,NROCED,CEN_EDU_f2,DISTRITO,LOCALIDAD_f2,...,P103C_SI_2,P103C_SI_3,P103C_SI_4,P103C_SI_A,P103C_SI_B,P104C_1,P104C_2,P104C_3,P104C_4,P104C_5,P104C_6,P104C_7,P104C_8,P104C_10,P104C_10_O,P105C,P105C_SI,P106C,P106C_4_OT,P107C,P108C,P108C_SI_1,P108C_SI_2,P108C_SI_3,P108C_SI_4,P108C_SI_5,P108C_SI_6,P108C_SI_7,P108C_SI_8,P109C,P109C_S_1,P109C_S_2,P109C_S_3,P109C_S_4,P109C_S_4O,P110C,P110C_S_1,P110C_S_2,P110C_S_3,P110C_S_3O,P111C,P112C_1,P112C_2,P112C_3,NIV_MOD_f2,GES_DEP_f2,AREA_CENSO_f2,CODGEO_f2,CODOOII_f2,_merge
0,,0000701,0,,DIVINO NIÑO JESUS,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,SURUPAMPA,SURUPAMPA,,,SURUPAMPA,2,Rural,040809,AREQUIPA,LA UNION,TAURIA,GRE AREQUIPA,040009,UGEL LA UNIÓN,0.000000,0.000000,4,PRONOEI,11,Mañana,,,,2,Inactivo,10-01-2012,05-08-2026,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1,,0000702,0,,SAN MARTIN THOURS,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,SAYLA,,0408080015,125203,SAYLA,2,Rural,040809,AREQUIPA,LA UNION,TAURIA,GRE AREQUIPA,040009,UGEL LA UNIÓN,-15.343310,-73.248575,4,PRONOEI,11,Mañana,,,,2,Inactivo,22-03-2012,05-08-2026,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
2,,0000704,0,,BAJO WAISIM,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,BAJO WAISIM,BAJO WAISIM,0104010146,654347,BAJO WAISIN,2,Rural,010401,AMAZONAS,CONDORCANQUI,NIEVA,DRE AMAZONAS,010004,UGEL INTERCULTURAL BILINGÜE CONDORCANQUI,-4.787293,-77.955081,15,Ciclo II - Entorno Comunitario,11,Mañana,,,,2,Inactivo,27-01-2016,05-08-2026,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
3,,0001505,0,,SANTA ROSA DE BELLAVISTA,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,BELLAVISTA S/N,,1510110015,121156,BELLAVISTA,2,Rural,151011,LIMA,YAUYOS,COLONIA,DRE LIMA PROVINCIAS,150206,UGEL 13 YAUYOS,-12.600308,-75.925076,4,PRONOEI,11,Mañana,,,,2,Inactivo,22-03-2012,05-08-2026,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
4,,0001506,0,,LOS LUCEROS,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,CARRETERA ACOBAMBILLA S/N,,1510110029,244062,BETHANIA,2,Rural,151011,LIMA,YAUYOS,COLONIA,DRE LIMA PROVINCIAS,150206,UGEL 13 YAUYOS,-12.637740,-75.533309,15,Ciclo II - Entorno Comunitario,11,Mañana,,,,2,Inactivo,22-03-2012,05-08-2026,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only


### <font color=green> 11. REVISAR EL RESULTADO DE LA INTEGRACIÓN </font>

In [ ]:

print("Resultado de la integración:")
display(base_integrada["_merge"].value_counts())

print("Valores faltantes en la base integrada:")
display(base_integrada.isnull().sum())

print("Resumen de la base integrada:")
display(base_integrada.describe(include="all"))

Resultado de la integración:


,count
_merge,
left_only,146695
both,34097
right_only,0


Valores faltantes en la base integrada:


,0
CODINST,0
COD_MOD,0
ANEXO,0
CODLOCAL_f1,0
CEN_EDU_f1,0
...,...
GES_DEP_f2,146695
AREA_CENSO_f2,146695
CODGEO_f2,146695
CODOOII_f2,146695


Resumen de la base integrada:


,CODINST,COD_MOD,ANEXO,CODLOCAL_f1,CEN_EDU_f1,NIV_MOD_f1,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,GESTION,D_GESTION,GES_DEP_f1,D_GES_DEP,DIRECTOR,TELEFONO,EMAIL,PAGWEB,DIR_CEN,LOCALIDAD_f1,CODCP_INEI,CODCCPP,CEN_POB,AREA_CENSO_f1,DAREACENSO,CODGEO_f1,D_DPTO,D_PROV,D_DIST,D_REGION,CODOOII_f1,D_DREUGEL,NLAT_IE,NLONG_IE,TIPOPROG,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,CODLOCAL_f2,NROCED,CEN_EDU_f2,DISTRITO,LOCALIDAD_f2,...,P103C_SI_2,P103C_SI_3,P103C_SI_4,P103C_SI_A,P103C_SI_B,P104C_1,P104C_2,P104C_3,P104C_4,P104C_5,P104C_6,P104C_7,P104C_8,P104C_10,P104C_10_O,P105C,P105C_SI,P106C,P106C_4_OT,P107C,P108C,P108C_SI_1,P108C_SI_2,P108C_SI_3,P108C_SI_4,P108C_SI_5,P108C_SI_6,P108C_SI_7,P108C_SI_8,P109C,P109C_S_1,P109C_S_2,P109C_S_3,P109C_S_4,P109C_S_4O,P110C,P110C_S_1,P110C_S_2,P110C_S_3,P110C_S_3O,P111C,P112C_1,P112C_2,P112C_3,NIV_MOD_f2,GES_DEP_f2,AREA_CENSO_f2,CODGEO_f2,CODOOII_f2,_merge
count,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792.000000,180792.000000,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,180792,34097,34097,34097,34097,34097,...,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,34097,180792
unique,68235,179383,26,105539,81916,29,21,3,3,3,3,3,10,10,80859,22302,6717,1083,86730,21059,32198,38955,30510,2,2,1892,25,196,1736,26,248,247,NaN,NaN,15,15,7,7,3594,4679,10138,2,2,13936,1,34091,1,19704,2795,16209,...,2,3,3,226,323,3,3,3,3,3,3,3,3,3,324,3,6,5,221,3,3,3,3,2,2,3,2,2,3,3,2,2,2,3,398,3,2,2,3,1111,3,3,3,3,3,10,2,1882,225,2
top,,1066018,0,,GOTITAS DE AMOR,A5,Inicial - Programa no escolarizado,Escolarizada,3,Mixto,1,Pública de gestión directa,A1,Sector Educación,,,,,PLAZA PRINCIPAL,,,523312,SAN JUAN DE LURIGANCHO,2,Rural,150132,LIMA,LIMA,SAN JUAN DE LURIGANCHO,DRE LIMA METROPOLITANA,150102,UGEL 01 SAN JUAN DE MIRAFLORES,NaN,NaN,,,11,Mañana,,,,1,Activo,,05-08-2026,825803,1A,INICIAL,SAN JUAN DE LURIGANCHO,,...,X,,,,,X,,,,,,,,,,SI,2,1,,NO,SI,,,,,,,,,SI,,,,,,SI,,,,,SI,,X,,A2,A1,2,150132,150102,left_only
freq,85232,26,179382,39696,527,55822,55822,120740,178734,178734,130758,130758,126405,126405,64401,143038,170937,178518,3038,97111,22568,2543,2671,92410,92410,2865,35180,29903,2865,29917,4899,4899,NaN,NaN,124970,124970,158226,158226,171816,170110,162420,115453,115453,21850,180792,2,34097,67,507,1300,...,17325,33639,33540,33637,33537,17132,25783,25022,26089,22607,29202,24966,25810,33626,33627,33525,18633,24983,33684,21762,21405,29043,24019,27438,26752,30015,27223,28483,32001,20630,20425,29383,21804,33382,33384,20840,20771,24814,31304,31304,26537,20345,18794,26360,32343,24593,18078,533,940,146695
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-10.411135,-75.379313,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.713466,7.054205,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N

### <font color=green> 12. PLANTILLA SIMPLE DE DICCIONARIO DE DATOS </font>

In [ ]:
# Completar interpretando sus variables.

diccionario_datos = pd.DataFrame({
    "variable": tabla_resumen_fuentes["variable"],
    "fuente": tabla_resumen_fuentes["fuente"],
    "descripcion": "",
    "tipo_general": "",       # Categórica o numérica
    "escala": "",             # Nominal, ordinal, intervalo o razón
    "rol_o_subtipo": "",      # Identificador, binaria, discreta, continua, fecha, texto, etc.
    "observaciones": ""
})

print("Plantilla de diccionario de datos:")
display(diccionario_datos)

Plantilla de diccionario de datos:


,variable,fuente,descripcion,tipo_general,escala,rol_o_subtipo,observaciones
0,CODINST,F1_Padron_web,,,,,
1,COD_MOD,F1_Padron_web,,,,,
2,ANEXO,F1_Padron_web,,,,,
3,CODLOCAL,F1_Padron_web,,,,,
4,CEN_EDU,F1_Padron_web,,,,,
...,...,...,...,...,...,...,...
278,NIVEL_EDUCATIVO,DenunciasEscolares,,,,,
279,TIPO_REPORTE,DenunciasEscolares,,,,,
280,TIPO_VIOLENCIA,DenunciasEscolares,,,,,
281,SUBTIPO_VIOLENCIA,DenunciasEscolares,,,,,


In [ ]:
# ===========================================================================
# PLANTILLA DE DICCIONARIO DE DATOS
# F1 (45 columnas) se documenta una por una a mano.
# F2 (230 columnas) se documenta por bloques y por patron de nomenclatura,
#     porque son las preguntas de la Cedula 1A del Censo Escolar.
# ===========================================================================
import re

# --- 1. F1: descripciones curadas -----------------------------------------
# formato: variable -> (descripcion, tipo_general, escala, rol_o_subtipo, observaciones)
DESC_F1 = {
 "CODINST":   ("Codigo de la institucion educativa (agrupa varios servicios)", "Categorica", "Nominal", "Identificador", "47.14% vacio; no usar como llave"),
 "COD_MOD":   ("Codigo modular del servicio educativo", "Categorica", "Nominal", "Identificador", "7 digitos con ceros a la izquierda; NO es unico (1,409 repetidos)"),
 "ANEXO":     ("Numero de anexo del servicio", "Categorica", "Nominal", "Identificador", "Completa a COD_MOD para formar la llave primaria real"),
 "CODLOCAL":  ("Codigo del local escolar donde funciona el servicio", "Categorica", "Nominal", "Identificador", "21.96% vacio; 69.78% vacio en no escolarizada -> sesgo"),
 "CEN_EDU":   ("Nombre del servicio educativo", "Categorica", "Nominal", "Texto", "Texto libre, 81,916 valores distintos"),
 "NIV_MOD":   ("Codigo de nivel y modalidad educativa", "Categorica", "Nominal", "Codigo", "29 codigos. A1 Cuna, A2 Jardin, A3 Cuna-jardin, A5 Inicial no escolarizado, B0 Primaria, F0 Secundaria"),
 "D_NIV_MOD": ("Descripcion del nivel y modalidad", "Categorica", "Nominal", "Descripcion de codigo", "21 descripciones para 29 codigos (L1-L9 colapsan en una sola)"),
 "D_FORMA":   ("Forma de atencion del servicio", "Categorica", "Nominal", "Descripcion de codigo", "Escolarizada / No escolarizada / No aplica"),
 "TIPSSEXO":  ("Codigo de sexo de los estudiantes atendidos", "Categorica", "Nominal", "Codigo", "1 Varones, 2 Mujeres, 3 Mixto"),
 "D_TIPSSEXO":("Descripcion del sexo atendido", "Categorica", "Nominal", "Descripcion de codigo", "99.0% Mixto: baja varianza"),
 "GESTION":   ("Codigo de tipo de gestion", "Categorica", "Nominal", "Codigo", "1 Publica gestion directa, 2 Publica gestion privada, 3 Privada"),
 "D_GESTION": ("Descripcion del tipo de gestion", "Categorica", "Nominal", "Descripcion de codigo", "Base de la variable derivada prop_publica"),
 "GES_DEP":   ("Codigo de dependencia administrativa", "Categorica", "Nominal", "Codigo", "10 codigos: A1-A4 publicos, B1-B6 privados"),
 "D_GES_DEP": ("Descripcion de la dependencia administrativa", "Categorica", "Nominal", "Descripcion de codigo", "Sector Educacion, Particular, Municipalidad, Comunidad, ..."),
 "DIRECTOR":  ("Nombre del director del servicio", "Categorica", "Nominal", "Texto", "35.62% vacio; dato personal, se excluye del modelo"),
 "TELEFONO":  ("Telefono de contacto", "Categorica", "Nominal", "Texto", "79.12% vacio; se descarta"),
 "EMAIL":     ("Correo electronico de contacto", "Categorica", "Nominal", "Texto", "94.55% vacio; se descarta"),
 "PAGWEB":    ("Pagina web del servicio", "Categorica", "Nominal", "Texto", "98.74% vacio; se descarta"),
 "DIR_CEN":   ("Direccion del local del servicio", "Categorica", "Nominal", "Texto", "Texto libre, no normalizado"),
 "LOCALIDAD": ("Localidad declarada", "Categorica", "Nominal", "Texto", "53.71% vacio"),
 "CODCP_INEI":("Codigo INEI del centro poblado", "Categorica", "Nominal", "Identificador", "12.48% vacio; llave hacia el directorio de CCPP del INEI"),
 "CODCCPP":   ("Codigo interno del centro poblado", "Categorica", "Nominal", "Identificador", "0.72% vacio"),
 "CEN_POB":   ("Nombre del centro poblado", "Categorica", "Nominal", "Texto", "30,510 valores; no normalizado"),
 "AREA_CENSO":("Codigo de area geografica", "Categorica", "Nominal binaria", "Binaria", "1 Urbana, 2 Rural"),
 "DAREACENSO":("Descripcion del area geografica", "Categorica", "Nominal binaria", "Binaria", "Base de la variable derivada prop_rural"),
 "CODGEO":    ("UBIGEO del distrito (2 dpto + 2 prov + 2 dist)", "Categorica", "Nominal", "Identificador", "LLAVE DE AGREGACION DEL MODELO. 1,892 distritos, siempre 6 digitos"),
 "D_DPTO":    ("Nombre del departamento", "Categorica", "Nominal", "Descripcion de codigo", "25 valores"),
 "D_PROV":    ("Nombre de la provincia", "Categorica", "Nominal", "Descripcion de codigo", "196 valores"),
 "D_DIST":    ("Nombre del distrito", "Categorica", "Nominal", "Descripcion de codigo", "1,736 nombres para 1,892 ubigeos: HAY HOMONIMOS, no unir por nombre"),
 "D_REGION":  ("Direccion Regional de Educacion", "Categorica", "Nominal", "Descripcion de codigo", "26 valores"),
 "CODOOII":   ("Codigo de la UGEL de pertenencia", "Categorica", "Nominal", "Identificador", "248 valores; corte administrativo alternativo"),
 "D_DREUGEL": ("Nombre de la DRE o UGEL", "Categorica", "Nominal", "Descripcion de codigo", "247 valores"),
 "NLAT_IE":   ("Latitud del servicio educativo", "Numerica", "Razon", "Continua", "1,302 registros en 0.000000 (nulo codificado); todos inactivos"),
 "NLONG_IE":  ("Longitud del servicio educativo", "Numerica", "Razon", "Continua", "1,302 registros en 0.000000 (nulo codificado); todos inactivos"),
 "TIPOPROG":  ("Codigo del tipo de programa no escolarizado", "Categorica", "Nominal", "Codigo", "69.12% vacio POR DISENO: solo aplica a NIV_MOD = A5"),
 "D_TIPOPROG":("Descripcion del tipo de programa no escolarizado", "Categorica", "Nominal", "Descripcion de codigo", "PRONOEI, SET, Ciclo I/II, PIETBAF, Ludoteca..."),
 "COD_TUR":   ("Codigo de turno de atencion", "Categorica", "Nominal", "Codigo", "11 Manana, 12 Tarde, 13 Manana-Tarde, 14 Noche, 15/16/17 combinaciones"),
 "D_COD_TUR": ("Descripcion del turno", "Categorica", "Nominal", "Descripcion de codigo", "87.5% Manana"),
 "NRORUC":    ("RUC del promotor (gestion privada)", "Categorica", "Nominal", "Identificador", "95.04% vacio POR DISENO: solo gestion privada"),
 "RZSOCIAL":  ("Razon social del promotor", "Categorica", "Nominal", "Texto", "94.09% vacio POR DISENO"),
 "PROMOTOR":  ("Nombre del promotor", "Categorica", "Nominal", "Texto", "89.84% vacio POR DISENO"),
 "ESTADO":    ("Codigo de situacion del servicio", "Categorica", "Nominal binaria", "Binaria", "1 Activo, 2 Inactivo"),
 "D_ESTADO":  ("Descripcion de la situacion del servicio", "Categorica", "Nominal binaria", "Binaria", "FILTRO DE INCLUSION: se trabaja con Activo (115,453)"),
 "FECHAREG":  ("Fecha de registro del servicio", "Categorica", "Intervalo (fecha como texto)", "Fecha", "12.09% vacio; contiene basura ('dd-mm-2007') y fechas imposibles (01-01-1899)"),
 "FECHA_ACT": ("Fecha de actualizacion del padron", "Categorica", "Intervalo (fecha como texto)", "Fecha", "CONSTANTE 05-08-2026: varianza cero, se descarta"),
}

# --- 2. F2: metadatos por bloque de la Cedula 1A ---------------------------
BLOQUES_F2 = {
 "A": "Bloque A - Funcionamiento del servicio, articulacion con salud y programas sociales, continuidad",
 "B": "Bloque B - Gestion institucional, participacion de las familias y convivencia escolar",
 "C": "Bloque C - Practica pedagogica, planificacion, acompanamiento y formacion docente",
}

IDENT_F2 = {
 "COD_MOD":   ("Codigo modular del servicio censado", "Identificador"),
 "ANEXO":     ("Numero de anexo", "Identificador"),
 "CODLOCAL":  ("Codigo del local escolar", "Identificador"),
 "NROCED":    ("Numero de cedula del Censo Escolar", "Identificador"),
 "CEN_EDU":   ("Nombre del servicio educativo", "Texto"),
 "DISTRITO":  ("Nombre del distrito declarado", "Texto"),
 "LOCALIDAD": ("Localidad declarada", "Texto"),
 "ALUM_35IIC":("Preliminar: atiende alumnos de 3 a 5 anos en IIEE de ciclo I", "Binaria"),
 "ALUM_02IC": ("Preliminar: atiende alumnos de 0 a 2 anos en ciclo I", "Binaria"),
 "PRO_CONVER":("Preliminar: el programa fue convertido a IIEE", "Binaria"),
 "ANIOCONVER":("Anio de conversion del programa a IIEE", "Fecha"),
 "NIV_MOD":   ("Codigo de nivel y modalidad", "Codigo"),
 "GES_DEP":   ("Codigo de dependencia administrativa", "Codigo"),
 "AREA_CENSO":("Codigo de area geografica", "Binaria"),
 "CODGEO":    ("UBIGEO del distrito", "Identificador"),
 "CODOOII":   ("Codigo de la UGEL", "Identificador"),
}

def clasificar_f2(col, serie):
    """Devuelve (descripcion, tipo_general, escala, rol_o_subtipo, observaciones)
    para una columna de la Cedula 1A, a partir de su nombre y de sus valores."""
    llenos  = serie[serie != ""]
    valores = set(llenos.unique())
    m = re.match(r"^P(\d{3})([ABC])(.*)$", col)
    if m is None:
        d, rol = IDENT_F2.get(col, ("Columna de identificacion", "Identificador"))
        return (d, "Categorica", "Nominal", rol, "Columna de identificacion o clasificacion")

    num, bloque, sufijo = m.group(1), m.group(2), m.group(3)
    preg = "P%s%s" % (num, bloque)
    ctx = "%s | pregunta %s de la cedula 1A" % (BLOQUES_F2[bloque], preg)

    # 1) Texto libre: se reconoce por la longitud de las respuestas, no por el sufijo
    #    (los sufijos reales son _OT, _5O, _S_6O, _10_O, _SI_A, _NO_O ... demasiado variados).
    if len(llenos) and llenos.str.len().max() > 20:
        return ("Respuesta abierta 'Otro, especifique' de %s" % preg,
                "Categorica", "Nominal", "Texto libre",
                ctx + ". Vacio = no eligio 'Otro'. Requiere normalizacion antes de usarse")

    if valores and valores <= {"SI", "NO", "NS"}:
        cond = bool(sufijo)
        return ("Respuesta a %s%s" % (preg, " (subpregunta %s)" % sufijo if cond else ""),
                "Categorica", "Nominal binaria", "Binaria SI/NO",
                ctx + (". Condicionada: vacio = no aplica (salto)" if cond
                       else ". Vacio = no respuesta real"))

    # 2) Casilla de marca: la proporcion de 'X' se mide SOBRE LOS VALORES LLENOS,
    #    porque muchas alternativas se marcan en menos del 1% de los servicios.
    if (valores and valores <= {"X", "0", "1", ".", "2"}
            and len(llenos) and (llenos == "X").mean() > 0.5):
        return ("Alternativa %s de la pregunta %s" % (sufijo.lstrip("_"), preg),
                "Categorica", "Nominal binaria", "Casilla de marca",
                ctx + ". X = marcada, VACIO = NO MARCADA (equivale a 0), NO es dato faltante. "
                      "Contiene valores contaminados: normalizar con (valor == 'X')")

    if valores and all(v.isdigit() for v in valores):
        return ("Cantidad o codigo reportado en %s%s" % (preg, sufijo),
                "Numerica" if len(valores) > 6 else "Categorica",
                "Razon" if len(valores) > 6 else "Nominal",
                "Conteo" if len(valores) > 6 else "Codigo",
                ctx + ". Almacenado como texto: convertir con pd.to_numeric(errors='coerce'). "
                      "Vacio = pregunta no aplicaba")

    return ("Respuesta a %s%s" % (preg, sufijo), "Categorica", "Nominal", "Por verificar",
            ctx + ". Patron de respuesta no estandar: revisar contra la cedula oficial")

# --- 3. Construccion del diccionario --------------------------------------
def construir_diccionario(datos, nombre_fuente, curado=None, clasificador=None):
    filas = []
    for col in datos.columns:
        serie = datos[col].astype(str).str.strip().replace({"None": "", "nan": ""})
        pct = round((serie == "").mean() * 100, 2)
        ejemplos = ", ".join(map(str, serie[serie != ""].value_counts().head(3).index.tolist()))[:80]

        if curado and col in curado:
            desc, tipo, escala, rol, obs = curado[col]
        elif clasificador is not None:
            desc, tipo, escala, rol, obs = clasificador(col, serie)
        else:
            desc, tipo, escala, rol, obs = ("", "", "", "", "")

        filas.append({
            "fuente": nombre_fuente,
            "variable": col,
            "descripcion": desc,
            "tipo_general": tipo,
            "escala": escala,
            "rol_o_subtipo": rol,
            "tipo_python": str(datos[col].dtype),
            "porcentaje_vacio": pct,
            "valores_distintos": int(serie[serie != ""].nunique()),
            "valores_frecuentes": ejemplos,
            "observaciones": obs,
        })
    return pd.DataFrame(filas)

dicc_f1 = construir_diccionario(df,  "F1_Padron_web",  curado=DESC_F1)
dicc_f2 = construir_diccionario(df1, "F2_Lineal_1A_1", clasificador=clasificar_f2)

diccionario_datos = pd.concat([dicc_f1, dicc_f2], ignore_index=True)

print("Diccionario de datos:", diccionario_datos.shape[0], "variables documentadas")
print("\nDistribucion de roles en F2 (Cedula 1A):")
display(dicc_f2["rol_o_subtipo"].value_counts())

print("\nF1 - Padron_web (45 variables):")
display(dicc_f1)

print("\nF2 - Lineal_1A_1 (muestra de cada tipo de patron):")
display(dicc_f2.groupby("rol_o_subtipo").head(2).sort_values("rol_o_subtipo"))

Diccionario de datos: 275 variables documentadas

Distribucion de roles en F2 (Cedula 1A):


,count
rol_o_subtipo,
Casilla de marca,119
Binaria SI/NO,54
Texto libre,21
Conteo,11
Codigo,9
Identificador,6
Binaria,4
Texto,3
Por verificar,2



F1 - Padron_web (45 variables):


,fuente,variable,descripcion,tipo_general,escala,rol_o_subtipo,tipo_python,porcentaje_vacio,valores_distintos,valores_frecuentes,observaciones
0,F1_Padron_web,CODINST,Codigo de la institucion educativa (agrupa var...,Categorica,Nominal,Identificador,object,47.14,68234,"25404092, 25875176, 22236261",47.14% vacio; no usar como llave
1,F1_Padron_web,COD_MOD,Codigo modular del servicio educativo,Categorica,Nominal,Identificador,object,0.00,179383,"1066018, 1114768, 0466425",7 digitos con ceros a la izquierda; NO es unic...
2,F1_Padron_web,ANEXO,Numero de anexo del servicio,Categorica,Nominal,Identificador,object,0.00,26,"0, 1, 2",Completa a COD_MOD para formar la llave primar...
3,F1_Padron_web,CODLOCAL,Codigo del local escolar donde funciona el ser...,Categorica,Nominal,Identificador,object,21.96,105538,"250758, 327282, 658446",21.96% vacio; 69.78% vacio en no escolarizada ...
4,F1_Padron_web,CEN_EDU,Nombre del servicio educativo,Categorica,Nominal,Texto,object,0.00,81916,"GOTITAS DE AMOR, LOS ANGELITOS, CORAZON DE JESUS","Texto libre, 81,916 valores distintos"
5,F1_Padron_web,NIV_MOD,Codigo de nivel y modalidad educativa,Categorica,Nominal,Codigo,object,0.00,29,"A5, B0, A2","29 codigos. A1 Cuna, A2 Jardin, A3 Cuna-jardin..."
6,F1_Padron_web,D_NIV_MOD,Descripcion del nivel y modalidad,Categorica,Nominal,Descripcion de codigo,object,0.00,21,"Inicial - Programa no escolarizado, Primaria, ...",21 descripciones para 29 codigos (L1-L9 colaps...
7,F1_Padron_web,D_FORMA,Forma de atencion del servicio,Categorica,Nominal,Descripcion de codigo,object,0.00,3,"Escolarizada, No escolarizada, No aplica",Escolarizada / No escolarizada / No aplica
8,F1_Padron_web,TIPSSEXO,Codigo de sexo de los estudiantes atendidos,Categorica,Nominal,Codigo,object,0.00,3,"3, 2, 1","1 Varones, 2 Mujeres, 3 Mixto"
9,F1_Padron_web,D_TIPSSEXO,Descripcion del sexo atendido,Categorica,Nominal,Descripcion de codigo,object,0.00,3,"Mixto, Mujeres, Varones",99.0% Mixto: baja varianza



F2 - Lineal_1A_1 (muestra de cada tipo de patron):


,fuente,variable,descripcion,tipo_general,escala,rol_o_subtipo,tipo_python,porcentaje_vacio,valores_distintos,valores_frecuentes,observaciones
7,F2_Lineal_1A_1,ALUM_35IIC,Preliminar: atiende alumnos de 3 a 5 anos en I...,Categorica,Nominal,Binaria,object,98.64,2,"SI, NO",Columna de identificacion o clasificacion
8,F2_Lineal_1A_1,ALUM_02IC,Preliminar: atiende alumnos de 0 a 2 anos en c...,Categorica,Nominal,Binaria,object,5.06,2,"NO, SI",Columna de identificacion o clasificacion
11,F2_Lineal_1A_1,P104A,Respuesta a P104A,Categorica,Nominal binaria,Binaria SI/NO,object,0.11,2,"NO, SI","Bloque A - Funcionamiento del servicio, articu..."
12,F2_Lineal_1A_1,P105A,Respuesta a P105A,Categorica,Nominal binaria,Binaria SI/NO,object,90.59,2,"SI, NO","Bloque A - Funcionamiento del servicio, articu..."
13,F2_Lineal_1A_1,P105A_S_1,Alternativa S_1 de la pregunta P105A,Categorica,Nominal binaria,Casilla de marca,object,97.24,1,X,"Bloque A - Funcionamiento del servicio, articu..."
14,F2_Lineal_1A_1,P105A_S_2,Alternativa S_2 de la pregunta P105A,Categorica,Nominal binaria,Casilla de marca,object,98.78,1,X,"Bloque A - Funcionamiento del servicio, articu..."
21,F2_Lineal_1A_1,P107A_NO,Cantidad o codigo reportado en P107A_NO,Categorica,Nominal,Codigo,object,95.79,4,"1, 4, 2","Bloque A - Funcionamiento del servicio, articu..."
144,F2_Lineal_1A_1,P120B,Cantidad o codigo reportado en P120B,Categorica,Nominal,Codigo,object,0.00,3,"1, 2, 3","Bloque B - Gestion institucional, participacio..."
63,F2_Lineal_1A_1,P119A_2_CP,Cantidad o codigo reportado en P119A_2_CP,Numerica,Razon,Conteo,object,89.29,31,"1, 2, 0","Bloque A - Funcionamiento del servicio, articu..."
62,F2_Lineal_1A_1,P119A_1_CP,Cantidad o codigo reportado en P119A_1_CP,Numerica,Razon,Conteo,object,92.05,25,"0, 1, 2","Bloque A - Funcionamiento del servicio, articu..."


### <font color=green> 13. EXPORTAR RESULTADOS </font>

In [ ]:
# --- Exportacion de resultados --------------------------------------------
ARCHIVO_SALIDA = "resultado_laboratorio_integracion.xlsx"

with pd.ExcelWriter(ARCHIVO_SALIDA) as writer:
    tabla_resumen_fuentes.to_excel(writer, sheet_name="resumen_fuentes",  index=False)
    diccionario_datos.to_excel(writer,     sheet_name="diccionario_datos", index=False)
    # La base integrada completa (180 mil filas x 273 columnas) excede el limite
    # practico de Excel: se exporta el subconjunto que realmente se analiza.
    inicial.to_excel(writer, sheet_name="inicial_integrada", index=False)

print("Archivo generado:", ARCHIVO_SALIDA)
print("  hoja resumen_fuentes   :", tabla_resumen_fuentes.shape)
print("  hoja diccionario_datos :", diccionario_datos.shape)
print("  hoja inicial_integrada :", inicial.shape)

Archivo generado: resultado_laboratorio_integracion.xlsx
  hoja resumen_fuentes   : (283, 6)
  hoja diccionario_datos : (275, 11)
  hoja inicial_integrada : (36127, 274)


# ROMPEBRECHAS
## Brechas territoriales en la oferta de Educación Inicial en el Perú

**Integrantes:** Mateo Pereyra · Raúl Porras · Mauro Martínez
**Curso:** Data Mining — 2026-II

---

## 1. Problema

El sistema educativo peruano distribuye sus servicios de forma muy desigual en el territorio. La oferta se concentra en zonas urbanas y capitales de provincia, mientras que amplias zonas rurales dependen de servicios no escolarizados o de instituciones con condiciones mínimas de funcionamiento. Esta desigualdad se discute normalmente en términos agregados (departamento o región) y no a nivel **distrital**, que es la unidad donde efectivamente se ejecuta la inversión pública y donde una familia decide si puede o no matricular a su hijo.

El proyecto se concentra en **Educación Inicial**, por tres razones:

1. Es el nivel donde la brecha es más severa: de los **53,650** servicios de inicial activos en el país, **17,523 (32.7%)** no son escolarizados (PRONOEI, SET, entornos comunitarios y familiares), es decir, no son un jardín con local e infraestructura propia.
2. Es el nivel más rural del sistema: **55.6%** de los servicios de inicial están en área rural, y en departamentos como Huancavelica (**93.7%**), Amazonas (**92.1%**) y Cajamarca (**87.3%**) la oferta es casi enteramente rural, frente a Callao (**0%**) y Lima (**8.6%**).
3. Es el único nivel para el que disponemos de una segunda fuente con información de condiciones de funcionamiento del servicio (la Cédula 1A del Censo Escolar).

**Pregunta de investigación:** ¿existen agrupamientos (*clusters*) de distritos con perfiles similares de oferta de Educación Inicial, y qué distritos concentran las mayores brechas según densidad de servicios, grado de escolarización, tipo de gestión, ruralidad y condiciones de funcionamiento del servicio?

**Objetivo general:** identificar y caracterizar perfiles distritales de oferta de Educación Inicial para detectar los distritos con déficit estructural.

**Objetivos específicos:**

- Construir indicadores distritales a partir del padrón: número de servicios de inicial activos, proporción no escolarizada, proporción rural, proporción pública, proporción con local escolar registrado.
- Añadir, para los distritos cubiertos por el Censo Escolar, indicadores de condiciones del servicio derivados de la Cédula 1A (materiales educativos, articulación con salud y programas sociales, participación de las familias, acompañamiento pedagógico, casos reportados de violencia).
- Segmentar los **1,886** distritos con oferta de inicial escolarizada activa mediante *clustering* no supervisado.
- Caracterizar los clusters y señalar cuáles corresponden a perfiles de déficit.

**Justificación:** el resultado permite priorizar dónde convertir programas no escolarizados en jardines, dónde reforzar el acompañamiento y dónde focalizar programas de acceso. La magnitud del problema se ve en un dato del propio padrón: la **mediana** de servicios de inicial escolarizada por distrito es de **9**, pero **312 distritos** (16.5%) tienen **2 o menos**, y **6 distritos** no tienen **ninguno**, mientras el máximo llega a **601** en un solo distrito de Lima.

> **Cambio de alcance respecto de la versión anterior de este documento.** El diseño original planteaba calcular ratios alumno/docente y alumno/sección a nivel nacional. El perfilado mostró que **ninguna de las fuentes disponibles contiene matrícula, docentes ni secciones**: `Padron_web.dbf` es un directorio administrativo (45 campos, ninguno de conteo) y `Lineal_1A_1.dbf` es un cuestionario de gestión, no un padrón de matrícula. Se reencuadró el problema a lo que las fuentes sí permiten medir. Los ratios quedan como línea de trabajo futura, condicionada a incorporar el módulo de matrícula del Censo Escolar.

## 2. Fuentes de datos a integrar

| # | Archivo | Fuente | Registros | Campos | Rol en el proyecto |
|---|---|---|---|---|---|
| **F1** | `Padron_web.dbf` | MINEDU – ESCALE, Padrón de Instituciones Educativas | **180,792** | **45** | **Base.** Universo de servicios educativos con ubicación, nivel, gestión, área, turno y estado |
| **F2** | `Lineal_1A_1.dbf` | MINEDU – ESCALE, Censo Escolar, **Cédula 1A** | **34,097** | **230** | **Enriquecimiento.** Condiciones de funcionamiento del servicio de Educación Inicial |
| F3 | *(pendiente)* | INEI – Proyecciones de población distrital | — | — | Denominador para convertir conteos en tasas |

**Enlaces:**

- Padrón de IIEE: https://escale.minedu.gob.pe/web/inicio/padron-de-iiee
- Censo Educativo: https://escale.minedu.gob.pe/censo-educativo

### Qué es realmente cada fuente

**F1 — `Padron_web.dbf`.** Directorio administrativo completo del sistema educativo peruano: un registro por servicio educativo (`COD_MOD` + `ANEXO`), de todos los niveles y modalidades, activos e inactivos. Es un archivo dBase III (versión `0x03`), última actualización **05-08-2026**. **No contiene ninguna variable de conteo** (ni alumnos, ni docentes, ni secciones): sus 45 campos son identificadores, clasificaciones, datos de contacto y coordenadas.

**F2 — `Lineal_1A_1.dbf`.** Es la **Cédula 1A del Censo Escolar** (el campo `NROCED` vale `'1A'` en las 34,097 filas). Aplica **únicamente a Educación Inicial escolarizada**: `NIV_MOD` toma solo los valores `A2` Jardín (32,343), `A3` Cuna-jardín (1,731) y `A1` Cuna (23). De sus 230 columnas, **16 son de identificación** y **214 son preguntas del cuestionario** (bloques A, B y C), casi todas categóricas: **54 preguntas binarias SI/NO**, **119 casillas de marca** (`X` cuando la alternativa fue seleccionada), **21 campos de texto libre** "Otro, especifique", **18 conteos o códigos** y 2 columnas con formato mixto. Última actualización **05-12-2025**.

> **Verificación de compatibilidad.** Los **34,097** pares `COD_MOD + ANEXO` de F2 existen **todos** en F1 (**100% de match**, 0 registros huérfanos). F2 cubre **34,015** de los **36,127** servicios de inicial escolarizada activos del padrón, es decir el **94.2%**, en **1,882** de los **1,892** distritos. Las **82** filas restantes corresponden a servicios que el padrón marca como inactivos: fueron censados en 2025 y dados de baja después. **Las dos fuentes son compatibles y la integración es sólida.**

## 3. Unidad de análisis

Se distinguen dos niveles, porque el registro base y el objeto del modelo no coinciden:

- **Unidad de observación (registro base):** el **servicio educativo**, identificado por `COD_MOD` + `ANEXO`. Un mismo local escolar puede albergar varios servicios (cuna, jardín, primaria), cada uno con su propio código modular. La verificación lo confirma: en F1 hay **1,409 `COD_MOD` repetidos**, pero **0 duplicados** de la combinación `COD_MOD + ANEXO`.
- **Unidad de análisis (para el modelo):** el **distrito**, identificado por `CODGEO` (UBIGEO de 6 dígitos). Es el nivel al que se agregan los indicadores y sobre el que se aplica el *clustering*.
- **Población objetivo:** los **1,886 distritos** con al menos un servicio de Educación Inicial escolarizada **activo**. El padrón registra 1,892 distritos en total; los 6 restantes se excluyen del modelo y se reportan aparte, porque un distrito sin ningún jardín es en sí mismo el caso extremo de brecha y no debe diluirse dentro de un cluster.

**Filtro de inclusión explícito.** Se trabaja con `D_ESTADO == 'Activo'` (**115,453** de 180,792 registros, el 63.9%). Los **65,339** registros inactivos se descartan del cálculo de indicadores pero se conservan en el perfilado, porque su distribución es informativa: el **100%** de las coordenadas en `(0,0)` corresponde a servicios inactivos.

> Formulación para la ficha: *"La unidad de observación es el servicio educativo de Educación Inicial (`COD_MOD`+`ANEXO`); la unidad de análisis para el modelamiento es el distrito (`CODGEO`), resultado de agregar los servicios activos por UBIGEO."*

## 4. Columnas comunes y llaves de integración

F1 y F2 comparten **8 columnas** con el mismo nombre: `COD_MOD`, `ANEXO`, `CODLOCAL`, `CEN_EDU`, `LOCALIDAD`, `NIV_MOD`, `GES_DEP`, `AREA_CENSO`, `CODGEO`, `CODOOII`. Solo dos de ellas sirven como llave.

| Llave | Formato | Une | Estado verificado |
|---|---|---|---|
| **`COD_MOD` + `ANEXO`** | 7 caracteres + 1 carácter | **F1 ↔ F2** | ✅ **Llave de integración registro a registro.** Única en ambas fuentes (0 duplicados). 34,097/34,097 filas de F2 encuentran par en F1 |
| **`CODGEO`** | 6 dígitos (2 dpto + 2 prov + 2 dist) | **F1 ↔ F2 ↔ F3** | ✅ **Llave de agregación del modelo.** 1,892 valores en F1, 1,882 en F2, todos de longitud 6, ninguno huérfano |
| `CODLOCAL` | 6 caracteres | F1 ↔ F1 (locales) | ⚠️ Usable, pero **vacío en el 21.96%** de F1. Relación 1:N con `COD_MOD` |
| `CODOOII` | 6 dígitos | F1 ↔ F2 (UGEL) | Corte administrativo alternativo: 248 UGEL en F1, 225 en F2 |
| `CODCP_INEI` / `CODCCPP` | 10 / 6 dígitos | F1 ↔ INEI CCPP | Solo si se baja a nivel de centro poblado. `CODCP_INEI` vacío en 12.48% |

### Precauciones críticas al integrar

- **`ANEXO` es constante `'0'` en F2** y toma 26 valores distintos en F1. Al unir, hay que construir la llave compuesta explícitamente (`COD_MOD + ANEXO`) y no confiar en que `COD_MOD` solo sea suficiente: los **138** servicios de inicial escolarizada activa con `ANEXO != '0'` se perderían o duplicarían.
- **Todo identificador debe leerse como texto.** `COD_MOD` (`'0415547'`), `CODGEO` (`'010101'` para Amazonas) y `CODLOCAL` tienen ceros a la izquierda. Leídos como entero, el join falla en silencio. Con `dbfread` todos los campos `C` llegan ya como `str`; el riesgo aparece al exportar a CSV y volver a leer.
- **Nunca unir por nombre de distrito.** `D_DIST` tiene **1,736** valores distintos para **1,892** ubigeos: hay homónimos entre departamentos. Además, `DISTRITO` en F2 tiene **2,795** valores distintos para 1,882 ubigeos, por variantes ortográficas y de tildes. El join debe ser siempre por `CODGEO`.

## 5. Tabla resumen de cada fuente

| Atributo | F1 Padrón IIEE | F2 Censo / Matrícula | F3 Población INEI |
|---|---|---|---|
| Formato | CSV / XLSX | CSV / XLSX | XLSX |
| N° de registros | ≈180 mil | ≈34 mil | **(por añadir)** |
| N° de columnas | =45 | 230 | **(por añadir)** |
| Granularidad | Servicio educativo | Servicio educativo × nivel | **(por añadir)** |
| Periodo | Corte 2025 | Censo Escolar 2025 | **(por añadir)** |
| Cobertura | Nacional | Nacional | Nacional |
| Llave | COD_MOD + ANEXO | COD_MOD + ANEXO | **(por añadir)** |
| Encoding | Latin-1 frecuente | Latin-1 | **(por añadir)** |
| Uso en el proyecto | Universo y atributos categóricos | Numerador de ratios | **(por añadir)** |

> Las cifras marcadas como **(por añadir)** deben completarse con el resultado de `df.shape` y del perfilado, no estimarse.

## 6. Diccionario de datos inicial

Los nombres exactos varían entre versiones del padrón; verificar con `df.columns`. Las columnas `D_*` son las descripciones textuales de un código.

| Variable | Tipo | Escala | Descripción | Valores esperados |
|---|---|---|---|---|
| `COD_MOD` | str | Nominal | Código modular del servicio | 7 dígitos |
| `ANEXO` | int | Nominal | Número de anexo | 0, 1, 2… |
| `CEN_EDU` | str | Nominal | Nombre de la IE | Texto libre |
| `NIV_MOD` / `D_NIV_MOD` | str | Nominal | Nivel y modalidad | Inicial, Primaria, Secundaria, EBA, EBE, ETP, Sup. Tecnológica, Sup. Pedagógica |
| `D_FORMA` | str | Nominal | Forma de atención | Escolarizada / No escolarizada |
| `D_GESTION` | str | Nominal | Tipo de gestión | Pública / Privada |
| `D_GES_DEP` | str | Nominal | Dependencia | Sector Educación, Otro sector público, Privada, Comunal, Convenio, Municipal |
| `CODLOCAL` | str | Nominal | Código de local escolar | 6 caracteres |
| `CODGEO` | str | Nominal | UBIGEO distrital | 6 dígitos |
| `D_DPTO` / `D_PROV` / `D_DIST` | str | Nominal | Nombres geográficos | Texto |
| `DAREACENSO` | str | Nominal | Área geográfica | Urbana / Rural |
| `D_COD_TUR` | str | Nominal | Turno | Mañana, Tarde, Noche, Continuo |
| `D_ESTADO` | str | Nominal | Situación del servicio | Activo, En receso, Clausurado, Cerrado temporal |
| `NLAT_IE` / `NLONG_IE` | float | Continua | Coordenadas geográficas | lat ≈ [-18.4, -0.03] · lon ≈ [-81.4, -68.6] |
| `TALUM_HOM` / `TALUM_MUJ` | int | Discreta | Alumnos por sexo | ≥ 0 |
| `TDOCENTE` | int | Discreta | Docentes | ≥ 0 |
| `TSECCION` | int | Discreta | Secciones | ≥ 0 |
| `CODOOII` | str | Nominal | UGEL de pertenencia | Código |

### Variables derivadas (nivel distrital — las que entran al clustering)

| Variable | Fórmula |
|---|---|
| `n_iiee_total`, `n_iiee_inicial`, `n_iiee_prim`, `n_iiee_sec` | Conteo de servicios activos por nivel |
| `iiee_por_1000hab` | n_iiee / población distrital × 1000 |
| `ratio_alum_doc` | Σ alumnos / Σ docentes |
| `ratio_alum_secc` | Σ alumnos / Σ secciones |
| `prop_rural` | IIEE en área rural / total |
| `prop_publica` | IIEE públicas / total |
| `prop_no_escolarizada` | Servicios no escolarizados / total |
| `cobertura_secundaria` | Matrícula 12–16 / población 12–16 |
| `flag_sin_secundaria` | 1 si el distrito no tiene ningún servicio de secundaria |

## 7. Problemas encontrados

### 7.1 Estructura e identificadores
- **Duplicados aparentes:** un mismo `COD_MOD` aparece varias veces con distinto `ANEXO`. Agrupar solo por COD_MOD subcuenta servicios.
- **Ceros a la izquierda destruidos** al abrir el CSV en Excel o al leer sin `dtype=str`: `COD_MOD`, `CODLOCAL` y `UBIGEO` se corrompen. Es el error más costoso porque rompe los joins de forma silenciosa.
- **Relación 1:N** entre local escolar y servicios: unir F1 con F3 sin cuidado infla los conteos.

### 7.2 Valores faltantes
- Coordenadas (`NLAT_IE`, `NLONG_IE`) ausentes o en `(0,0)`, típicamente en zonas rurales — justo donde está la brecha, lo que genera **sesgo de selección** si se descartan esas filas.
- `TDOCENTE`, `TALUMNO`, `TSECCION` vacíos en servicios no escolarizados (PRONOEI) y en parte de la gestión privada, que reporta con menor tasa de respuesta.
- **Ambigüedad entre `0` real y nulo codificado como 0:** un colegio con 0 docentes es probablemente dato faltante, no un colegio sin docentes.

### 7.3 Tipos mal detectados
- `UBIGEO` leído como `int64` (Amazonas `010101` → `10101`).
- Conteos numéricos leídos como `object` por separadores de miles o espacios.
- Fechas de registro como texto en formatos mixtos.
- **Encoding Latin-1 no declarado:** `Áncash` → `A?ncash`, lo que rompe cualquier join por nombre de distrito.

### 7.4 Categorías raras o inconsistentes
- Categorías de baja frecuencia en `D_NIV_MOD` (EBE, ETP, Superior Pedagógica) no comparables con EBR; conviene separarlas o excluirlas del cálculo de ratios.
- Variantes ortográficas y de acentuación en nombres de distrito (`Nasca`/`Nazca`, con y sin tildes) — razón adicional para unir por UBIGEO y **nunca** por nombre.
- `D_ESTADO` con múltiples valores fuera de "Activo": hay que definir explícitamente el criterio de inclusión.
- Distritos de creación reciente presentes en el padrón MINEDU pero ausentes en las proyecciones INEI, o viceversa → filas huérfanas en el join.

### 7.5 Inconsistencias lógicas y valores extremos
- `ratio_alum_doc` infinito o indefinido cuando `TDOCENTE = 0`.
- Ratios extremos (> 60 alumnos/docente) que pueden ser reales en zonas críticas o errores de captura: revisarlos manualmente antes de winsorizar, porque son precisamente los casos de interés.
- Servicios cuyo `UBIGEO` no coincide con el departamento declarado en `D_DPTO`.
- Distritos con muy pocos servicios, donde los promedios son inestables.

### 7.6 Sesgos y limitaciones del diseño
- El padrón mide **oferta registrada**, no acceso efectivo: un distrito con muchas IIEE puede seguir teniendo baja asistencia.
- Sin la fuente de población (F3), la densidad de IIEE no es un indicador de brecha sino de **tamaño del distrito**.
- No hay variable de distancia ni tiempo de traslado, que es el determinante real del acceso en zonas rurales andinas y amazónicas.
- Las variables numéricas están en escalas muy distintas → **estandarización obligatoria** antes de K-means; evaluar K-prototypes si se mantienen variables categóricas.

## 8. Decisiones de limpieza justificadas

Se documentan las decisiones tomadas sobre las tres fuentes, priorizando F4 (denuncias escolares)

### 8.1 Qué se corrigió
- **Formato de texto de `UGEL`** (F4): se normalizó a un único patrón (`"NN Nombre"`, mayúsculas, sin tildes, sin doble espacio), eliminando el guion en las 10 UGEL que lo traían, para que el campo sea comparable consigo mismo y con otras fuentes.
- **Prefijo `DRE`/`GRE`** (F1 vs F4): se normalizó `GRE` → `DRE` en `D_DREUGEL` del padrón antes de comparar contra `denuncias_escolares`, porque ambos prefijos designan la misma entidad administrativa (Dirección/Gerencia Regional de Educación); mantenerlos distintos rompería el join en 9 regiones.
- **Acentos y mayúsculas** en `DRE`/`UGEL` (F4) y `D_DREUGEL` (F1): se normalizaron a un mismo estándar (mayúsculas, sin tildes) exclusivamente para efectos de *matching*; se conserva la columna original sin tocar para reporte y visualización.

### 8.2 Qué se imputó
- **No se imputó ningún valor en F4.** No hay celdas vacías (0% de faltantes en las 8 columnas), así que no hay nada que imputar.
- En F1/F2, la imputación sigue limitada a los casos ya descritos en la sección 7.2 (coordenadas en `(0,0)`, docentes/secciones vacíos en no escolarizados): **se decidió no imputar coordenadas ni conteos**, y en su lugar se dejará un *flag* booleano (`coord_valida`, `dato_reportado`) porque imputar la ubicación o el conteo de un servicio inventaría información que no existe en la fuente.

### 8.3 Qué se eliminó
- **No se eliminaron las 7,282 filas duplicadas exactas de F4.** Al no existir un identificador de caso (`ID_CASO`), eliminar duplicados asumiría —sin evidencia— que son el mismo caso repetido, cuando pueden ser casos distintos con el mismo perfil categórico (misma fecha, misma UGEL, mismo tipo de violencia). Eliminar reportes reales de violencia escolar es el error más costoso posible en esta base: se prefiere sobreconteo documentado a subconteo silencioso.
- Si el objetivo del análisis exige un conteo conservador (por ejemplo, para comparar tasas entre UGEL), se calculará también un conteo alternativo con duplicados exactos colapsados a 1, y **se reportarán ambos números lado a lado**, nunca solo uno.
- Se mantiene la decisión ya tomada en F1: no eliminar los 65,339 registros con `D_ESTADO != 'Activo'`; se excluyen del cálculo de indicadores pero se conservan en el dataframe base.

### 8.4 Qué se mantuvo y por qué
- Se mantuvo el registro `UGEL == "DRE Callao"` (1,328 filas) tal como viene: no es un error, Callao no se subdivide en UGEL y la propia DRE cumple esa función administrativa, lo cual es coherente con que `D_DREUGEL` en el padrón también use `"DRE CALLAO"` para esos colegios.
- Se mantuvo la categoría `"Con uso de armas"` clasificada tanto en `Física` como `Psicológica`: no se fuerza a una sola categoría porque la ambigüedad refleja un criterio real del formulario SíseVe (amenaza con arma sin contacto físico vs. agresión con arma), no un error de captura.
- Se mantuvieron las 4 UGEL de F4 que no encuentran par exacto en F1 tras la normalización (`UGEL 03 Cercado`, `UGEL Antonio Raimondi`, `UGEL Carlos F. Fitzcarrald`, `UGEL Pichari‑Kimbiri‑Villa Virgen`) sin forzar un match aproximado automático: se documentan como pendientes de un diccionario de equivalencias manual (ver sección 9), porque un *fuzzy match* automático sobre nombres administrativos puede unir UGEL distintas silenciosamente.


In [ ]:

# 8. IMPLEMENTACIÓN DE LAS DECISIONES DE LIMPIEZA

import unicodedata

def normaliza_texto(s):
    """Mayusculas, sin tildes, sin guion, espacios simples. Solo para matching."""
    s = str(s).upper().strip()
    s = s.replace("-", " ")
    s = " ".join(s.split())
    s = "".join(c for c in unicodedata.normalize("NFD", s) if unicodedata.category(c) != "Mn")
    return s

# Copia de trabajo: se conservan DRE/UGEL originales, se agregan versiones normalizadas
df2_limpio = df2.copy()
df2_limpio["DRE_norm"] = df2_limpio["DRE"].apply(normaliza_texto)
df2_limpio["UGEL_norm"] = df2_limpio["UGEL"].apply(normaliza_texto)

# Se flaggean duplicados exactos SIN eliminarlos (decision documentada en 8.3)
df2_limpio["es_duplicado_exacto"] = df2_limpio.duplicated(keep=False)
print("Duplicados marcados (no eliminados):", df2_limpio["es_duplicado_exacto"].sum())

# Conteo alternativo (conservador), reportado aparte, nunca reemplazando al original
n_casos_reportados = len(df2_limpio)
n_casos_sin_duplicado_exacto = len(df2_limpio.drop_duplicates())
print(f"Conteo de casos, versión completa: {n_casos_reportados}")
print(f"Conteo de casos, colapsando duplicados exactos: {n_casos_sin_duplicado_exacto}")

display(df2_limpio.head())


In [ ]:

# 8.y TABLA DE REGISTRO DE DECISIONES DE LIMPIEZA
decisiones = pd.DataFrame([
    ["UGEL", "Categorías no uniformes (con/sin guion: '01 - El Porvenir' vs '01 Comas')",
     "Normalizar a mayúsculas, sin tildes, sin guion (UGEL_norm)",
     "Sin esto, un merge por texto exacto contra otra fuente falla en 10/225 UGEL sin dar error."],
    ["D_DREUGEL (F1)", "Prefijo inconsistente entre fuentes: 'GRE' en F1 vs 'DRE' en F4 para 9 regiones",
     "Reemplazar 'GRE ' por 'DRE ' antes de comparar",
     "'GRE Arequipa' y 'DRE Arequipa' son la misma entidad; mantenerlos distintos rompe el join en esas 9 regiones."],
    ["UGEL vs DRE", "Ambigüedad de llave: 'UGEL La Unión' existe en Piura y en Arequipa",
     "Usar la llave compuesta DRE + UGEL, nunca UGEL sola",
     "Usar solo UGEL fusionaría en un mismo grupo casos de dos regiones distintas."],
    ["fila duplicada (F4)", "7,282 filas duplicadas exactas (25% de las 48,273) sin ID de caso",
     "Se flaggean con 'es_duplicado_exacto'; NO se eliminan",
     "Al no existir ID_CASO, borrar duplicados asumiría sin evidencia que son el mismo caso; el costo de subcontar violencia escolar es mayor que el de un posible sobreconteo."],
    ["SUBTIPO_VIOLENCIA", "'Con uso de armas' aparece bajo Física (51) y Psicológica (19)",
     "Se mantiene la doble clasificación tal como viene",
     "Refleja un criterio real del formulario SíseVe (amenaza vs. agresión con arma), no un error de captura."],
    ["NLAT_IE / NLONG_IE (F1)", "Coordenadas en (0,0), concentradas en servicios inactivos y zona rural",
     "No se imputa la ubicación; se deja como faltante y se agrega flag 'coord_valida'",
     "Imputar una coordenada inventaría una ubicación que la fuente no reporta, y el patrón de faltante es informativo (sesgo geográfico) y no debe ocultarse."],
    ["D_ESTADO (F1)", "65,339 registros con estado distinto de 'Activo' (36.1%)",
     "No se eliminan del dataframe; se excluyen solo del cálculo de indicadores",
     "Un registro inactivo sigue siendo evidencia útil para el perfilado (p. ej. el 100% de coordenadas (0,0) son de servicios inactivos)."],
], columns=["variable", "problema", "decisión", "justificación"])

pd.set_option("display.max_colwidth", None)
display(decisiones)


## 9. Base integrada inicial: incorporación de F4 (denuncias escolares)

F4 tiene una **granularidad distinta** a F1/F2 (caso reportado vs. servicio educativo) y **no comparte ninguna llave exacta tipo `COD_MOD`**: solo trae texto libre de DRE/UGEL. Por eso no puede integrarse fila a fila como F1↔F2; el plan es una **integración por agregación** a nivel UGEL.

**Plan de integración:**
1. Normalizar `D_DREUGEL` (F1) y `DRE`+`UGEL` (F4) con la misma función (`normaliza_texto`): mayúsculas, sin tildes, sin guiones, prefijo `GRE`→`DRE`.
2. Agregar F4: número de casos reportados por `UGEL_norm` (y opcionalmente por `TIPO_VIOLENCIA`).
3. Agregar F1: número de servicios activos por `D_DREUGEL_norm` (mismo campo usado como llave de UGEL en el padrón).
4. Unir ambas tablas agregadas por la llave de texto normalizada (`UGEL_norm`), **no** al nivel de `COD_MOD` porque F4 no lo trae.
5. Reportar explícitamente el **porcentaje de UGEL y de filas que no encuentran par**, en vez de descartarlas en silencio.

Se probó el paso 1–4 para medir qué tan viable es la llave de texto:


In [ ]:

# 9 PRUEBA DE INTEGRACIÓN F1 <-> F4 POR UGEL NORMALIZADA

# UGEL normalizadas presentes en cada fuente
ugel_f4 = set(df2_limpio["UGEL_norm"].unique())

def normaliza_dreugel_f1(s):
    s = normaliza_texto(s)
    s = s.replace("GRE ", "DRE ", 1) if s.startswith("GRE ") else s
    return s

ugel_f1 = set(normaliza_dreugel_f1(rec["D_DREUGEL"]) for rec in tabla_f1)

interseccion = ugel_f4 & ugel_f1
sin_match = ugel_f4 - ugel_f1

print(f"UGEL distintas en F4: {len(ugel_f4)}")
print(f"UGEL distintas en F1 (D_DREUGEL normalizado): {len(ugel_f1)}")
print(f"UGEL de F4 que encuentran par en F1: {len(interseccion)} ({len(interseccion)/len(ugel_f4)*100:.1f}%)")
print(f"UGEL de F4 SIN par en F1: {sorted(sin_match)}")

filas_sin_match = (~df2_limpio["UGEL_norm"].isin(ugel_f1)).sum()
print(f"\nFilas de F4 cubiertas por el join: {len(df2_limpio) - filas_sin_match} "
      f"({(len(df2_limpio)-filas_sin_match)/len(df2_limpio)*100:.1f}%)")
print(f"Filas de F4 SIN cobertura (quedan fuera del análisis integrado): {filas_sin_match} "
      f"({filas_sin_match/len(df2_limpio)*100:.1f}%)")

# --- Base integrada inicial a nivel UGEL
casos_por_ugel = (
    df2_limpio.groupby("UGEL_norm")
    .size()
    .rename("n_casos_violencia")
    .reset_index()
)

servicios_f1 = pd.DataFrame(iter(tabla_f1))
servicios_f1["UGEL_norm"] = servicios_f1["D_DREUGEL"].apply(normaliza_dreugel_f1)
servicios_activos_por_ugel = (
    servicios_f1[servicios_f1["D_ESTADO"] == "Activo"]
    .groupby("UGEL_norm")
    .size()
    .rename("n_servicios_activos")
    .reset_index()
)

base_integrada_ugel = pd.merge(
    servicios_activos_por_ugel, casos_por_ugel, on="UGEL_norm", how="outer", indicator=True
)
print("\nResultado del merge (base integrada a nivel UGEL):")
display(base_integrada_ugel["_merge"].value_counts())
display(base_integrada_ugel.head())
